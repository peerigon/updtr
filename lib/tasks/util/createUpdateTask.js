"use strict";

function createUpdateTask(instance, outdated) {
    return {
        name: outdated.name,
        type: outdated.type,
        updateTo: instance.wanted ? outdated.wanted : outdated.latest,
        rollbackTo: outdated.current,
    };
}

module.exports = createUpdateTask;
