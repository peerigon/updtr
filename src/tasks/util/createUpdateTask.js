import semver from "semver";
import {
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
} from "../../constants/config";

function determineUpdateToVersion({ updateTo }, { current, wanted, latest }) {
    switch (updateTo) {
        case UPDATE_TO_LATEST:
            return latest;
        case UPDATE_TO_WANTED:
            return wanted;
        case UPDATE_TO_NON_BREAKING:
    }

    const range = "^" + current;

    if (semver.satisfies(latest, range) === true) {
        return latest;
    }
    if (semver.satisfies(wanted, range) === true) {
        return wanted;
    }

    return current;
}

export default function createUpdateTask(outdated, updtrConfig) {
    return {
        name: outdated.name,
        updateTo: determineUpdateToVersion(updtrConfig, outdated),
        rollbackTo: outdated.current,
    };
}
