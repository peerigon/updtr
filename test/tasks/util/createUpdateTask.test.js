import semver from "semver";
import createUpdateTask from "../../../src/tasks/util/createUpdateTask";
import {
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
} from "../../../src/constants/config";
import FakeUpdtr from "../../helpers/FakeUpdtr";
import outdateds from "../../fixtures/outdateds";

function stringify(outdated) {
    return [
        "current " + outdated.current,
        "wanted " + outdated.wanted,
        "latest " + outdated.latest,
    ].join(", ");
}

expect.extend({
    toSatisfySemver(received, argument) {
        const pass = semver.satisfies(received, argument);

        if (pass === true) {
            return {
                message: () =>
                    `expected ${ received } not to satisfy ${ argument }`,
                pass,
            };
        }

        return {
            message: () => `expected ${ received } to satisfy ${ argument }`,
            pass,
        };
    },
});

describe("createUpdateTask()", () => {
    test("should return a valid update task", () => {
        expect(
            createUpdateTask(outdateds[0], {
                ...FakeUpdtr.baseConfig,
                updateTo: UPDATE_TO_LATEST,
            })
        ).toMatchSnapshot();
    });
    describe(`when updateTo is "${ UPDATE_TO_LATEST }"`, () => {
        outdateds.forEach(outdated => {
            test(`when given ${ stringify(outdated) } should update to ${ outdated.latest } and rollback to ${ outdated.current }`, () => {
                const updateTask = createUpdateTask(outdated, {
                    ...FakeUpdtr.baseConfig,
                    updateTo: UPDATE_TO_LATEST,
                });

                expect(updateTask.updateTo).toBe(outdated.latest);
                expect(updateTask.rollbackTo).toBe(outdated.current);
            });
        });
    });
    describe(`when updateTo is "${ UPDATE_TO_NON_BREAKING }"`, () => {
        outdateds.forEach(outdated => {
            const range = "^" + outdated.current;

            test(`when given ${ stringify(outdated) } should return an update that satisfies ${ range } and rollback to ${ outdated.current }`, () => {
                const updateTask = createUpdateTask(outdated, {
                    ...FakeUpdtr.baseConfig,
                    updateTo: UPDATE_TO_NON_BREAKING,
                });

                expect(updateTask.updateTo).toSatisfySemver(range);
                expect(updateTask.rollbackTo).toBe(outdated.current);
            });
        });
    });
    describe(`when updateTo is "${ UPDATE_TO_WANTED }"`, () => {
        outdateds.forEach(outdated => {
            test(`when given ${ stringify(outdated) } should update to ${ outdated.wanted } and rollback to ${ outdated.current }`, () => {
                const updateTask = createUpdateTask(outdated, {
                    ...FakeUpdtr.baseConfig,
                    updateTo: UPDATE_TO_WANTED,
                });

                expect(updateTask.updateTo).toBe(outdated.wanted);
                expect(updateTask.rollbackTo).toBe(outdated.current);
            });
        });
    });
});
