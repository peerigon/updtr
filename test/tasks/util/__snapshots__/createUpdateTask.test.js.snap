// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUpdateTask() should return a valid update task 1`] = `
Object {
  "name": "module",
  "rollbackTo": "0.0.0",
  "updateTo": "0.0.1",
}
`;
